lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -674,83 +843,446 @@ export default function getAPIImplementation(): APIInterface {
-                {},
-                {},
-                userContext
-            );
-            return {
-                status: "OK",
-                session,
-                user: response.user,
-            };
-        },
-    };
-}
+        if (!foundVerifiedEmail) {
+            throw new SignUpError({
+                status: "SIGN_UP_NOT_ALLOWED",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_010)",
+            });
+        }
+    }
+};
+
+const checkIfFactorUserBeingCreatedCanBeLinkedWithSessionUser = async (
+    tenantId: string,
+    sessionUser: User,
+    accountInfo: { email: string },
+    userContext: UserContext
+) => {
+    if (!sessionUser.isPrimaryUser) {
+        const canCreatePrimary = await AccountLinking.getInstance().recipeInterfaceImpl.canCreatePrimaryUser({
+            recipeUserId: sessionUser.loginMethods[0].recipeUserId,
+            userContext,
+        });
+
+        if (canCreatePrimary.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+            // Session user is linked to another primary user, which means the session is revoked as well
+            throw new SessionError({
+                type: SessionError.TRY_REFRESH_TOKEN,
+                message: "Session may be revoked",
+            });
+        }
+
+        if (canCreatePrimary.status === "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR") {
+            throw new SignUpError({
+                status: "SIGN_UP_NOT_ALLOWED",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_011)",
+            });
+        }
+    }
+
+    // Check if the linking with session user going to fail and avoid user creation here
+    const users = await listUsersByAccountInfo(tenantId, accountInfo, true, userContext);
+    for (const user of users) {
+        if (user.isPrimaryUser && user.id !== sessionUser.id) {
+            throw new SignUpError({
+                status: "SIGN_UP_NOT_ALLOWED",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_012)",
+            });
+        }
+    }
+};
+
+const linkAccountsForFactorSetup = async (sessionUser: User, recipeUserId: RecipeUserId, userContext: UserContext) => {
+    if (!sessionUser.isPrimaryUser) {
+        const createPrimaryRes = await AccountLinking.getInstance().recipeInterfaceImpl.createPrimaryUser({
+            recipeUserId: new RecipeUserId(sessionUser.id),
+            userContext,
+        });
+        if (createPrimaryRes.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+            // Session user is linked to another primary user, which means the session is revoked as well
+            throw new SessionError({
+                type: SessionError.TRY_REFRESH_TOKEN,
+                message: "Session may be revoked",
+            });
```

Comment:
throw new recurse error

Reply:
Shuold be unauthorised error, not recursion error

Reply:
Done